Check that castToNonNull method is not passed @NonNull

lazaroclapp opened the pull request:
This adds a NullAway warning whenever all the following conditions apply:

- A `castToNonNull` method is configured with `-XepOpt:NullAway:CastToNonNullMethod=...`
- This method is passed an argument that NullAway knows to be `@NonNull`
- The invocation happens outside an initializer (inside an initializer, an `@NonNull` field might be passed to `castToNonNull` to "initialize it" in some weird cases where the actual reason for initialization is not detected by NullAway)

(See #190)

**Important**: This is a breaking change that adds a new `WARNING` level alert on a case that a) is safe, b) we previously allowed, but which I believe we should disallow for code quality purposes. We could gate this by a config flag, but I tentatively prefer not to proliferate those, and I can't see a good reason to need to pass `@NonNull` to `castToNonNull` other than refactoring issues and code written for earlier less precise versions of NullAway. I tentatively vote for a version bump to 0.5.0 instead of a config flag, but can do either.

lazaroclapp commented:
ToDo: Add fix suggestion code.

msridhar commented:
@lazaroclapp is this ready for review or still WIP?

lazaroclapp commented:
@msridhar Ready to review, plus ready for your opinion on whether it needs a config flag or just a bump to `0.5.0`. Not super urgent though, but certainly ready.

msridhar commented on nullaway/src/main/java/com/uber/nullaway/NullAway.java:
```diff
@@ -1152,20 +1154,54 @@ private Description handleInvocation(
             MessageTypes.PASS_NULLABLE, actual, message, actual, state.getPath());
       }
     }
+    // Check for @NonNull being passed to castToNonNull (if configured)
```
Can we extract this logic to its own function?

lazaroclapp replied:
Yup.

jamey713 commented:
Its working on my phone, I've had someone deliberately, hacking every phone I've had, since October 2017, this is actually, working , thank you

msridhar commented:
Oh, and I agree with bumping to 0.5.0 rather than adding another config flag

lazaroclapp commented:
@msridhar I addressed the comment and added some basic auto-fix tests. Let me know if you think this is ready to land and I'll cut 0.5.0 right after unless you have any other changes that want to get into that version.
